File: michielstokman/ai-implementation-fastapi/app/providers/openai_provider.py
import json
import re
import httpx
from app.config import settings


async def generate_json_from_prompt(prompt: str) -> dict:
    if not settings.openai_api_key:
        raise ValueError('OPENAI_API_KEY is required when USE_MOCK_AI=false')

    payload = {
        'model': settings.ai_model,
        'temperature': 0.7,
        'messages': [
            {'role': 'system', 'content': 'Return valid JSON only.'},
            {'role': 'user', 'content': prompt},
        ],
    }

    async with httpx.AsyncClient(timeout=45.0) as client:
        res = await client.post(
            'https://api.openai.com/v1/chat/completions',
            headers={
                'Authorization': f'Bearer {settings.openai_api_key}',
                'Content-Type': 'application/json',
            },
            json=payload,
        )

    if res.status_code >= 400:
        raise RuntimeError(f'OpenAI request failed: {res.status_code} {res.text}')

    text = res.json().get('choices', [{}])[0].get('message', {}).get('content', '{}')

    try:
        return json.loads(text)
    except json.JSONDecodeError:
        match = re.search(r'\{.*\}', text, re.S)
        if not match:
            raise ValueError('Model response was not valid JSON')
        return json.loads(match.group(0))

File: michielstokman/ai-implementation/src/providers/openaiProvider.js
const { config } = require("../config");

async function generateJsonFromPrompt(prompt) {
  if (!config.openAiApiKey) {
    throw new Error("OPENAI_API_KEY is required when USE_MOCK_AI=false");
  }

  const res = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${config.openAiApiKey}`,
    },
    body: JSON.stringify({
      model: config.model,
      temperature: 0.7,
      messages: [
        { role: "system", content: "Return valid JSON only." },
        { role: "user", content: prompt },
      ],
    }),
  });

  if (!res.ok) {
    throw new Error(`OpenAI request failed: ${res.status} ${await res.text()}`);
  }

  const data = await res.json();
  const text = data?.choices?.[0]?.message?.content || "{}";

  try {
    return JSON.parse(text);
  } catch {
    const start = text.indexOf("{");
    const end = text.lastIndexOf("}");
    if (start >= 0 && end > start) {
      return JSON.parse(text.slice(start, end + 1));
    }
    throw new Error("Model response was not valid JSON");
  }
}

module.exports = { generateJsonFromPrompt };
